[DM-33981] Update build system and packaging

rra opened the pull request:
- Unpin all dependencies to avoid an incompatibility between Jinja2 and markupsafe
- Enable black and standard pre-commit configuration
- Run black on all source code
- Reorganize the source to use a `src` subdirectory
- Rewrite the build system to use `pyproject.toml` and `setup.cfg` with setuptools_scm
- Use `tox` for testing and documentation builds
- Add type annotations and run mypy as part of the build
- Replace TravisCI with GitHub Actions
- Use `importlib.metadata` instead of `pkg_resources` for the module version
- Update schema for the latest Cerberus, which renames two parameters

rra committed:
Uncap dependency versions, bump test dependencies

For all install_requires, remove the major version pins.  This
concretely fixes an incompatibility between markupsafe and the
old 2.x version of Jinja2, and more generally matches our
practices for other library packages.

rra committed:
Run black on all source code

rra committed:
Move source to src directory

Follow the new package layout for SQuaRE packages.

rra committed:
Rework build system

Use pyproject.toml and setup.cfg following the new model for
SQuaRE projects, and turn setup.py into a small stub.

rra committed:
Add pre-commit configuration

rra committed:
Add tox configuration

rra committed:
Add Makefile to install pre-commit and tox

rra committed:
Replace Travis CI with GitHub Actions

rra committed:
Add __all__ to top-level package

Declare TemplatekitExtension in __all__ to suppress a lint warning.

rra committed:
Wrap too-long lines

Two docstrings go over 80 characters after black reformatting.
Wrap them.

rra committed:
Fix mypy configuration

Exclude tests/data, which looks like Python but isn't, and remove
the stray sqlalchemy plugin configuration that was accidentally
copied from Safir.

rra committed:
Use importlib.metadata for module version

Obtain the module version using importlib.metadata instead of
pkg_resources, and export a version_info variable as well matching
how Safir works.  This gets rid of mypy warnings for using
setuptools modules without type stubs.

rra committed:
Remove asyncio_mode configuration

We're not using pytest-asyncio for this package (at least yet).

rra committed:
Add type annotations

Add type annotations to all source code so that mypy checks will
pass.

rra committed:
Update schema for latest Cerberus

Cerberus has renamed keyschema to keysrules and valueschema to
valuesrules.  Update the schema accordingly.

rra committed:
Add changelog for 0.5.0 release

rra committed:
Update documentation and its build

Replace the development documentation with documentation taken from
Safir, with some modifications for submodules.  Remove an object
reference from the changelog since we're not building API
documentation.  Use a tox environment to do the documentation
build.

rra committed:
Add type checking to the CI build

rra committed:
Enable submodules for CI

Check out the submodules so that the test suite data is present.

jonathansick commented on docs/index.rst:
```diff
@@ -30,11 +39,11 @@ Template development guide
 
-   changelog
-   development-guide
+   dev/development
+   dev/release
```
I'd kind of prefer that we didn't break/change URLs in a Python 3.10 support PR just for the sake of consistency. But at the end of the day, this is fine.

rra replied:
I can merge those back into one page, or drop the release page, if you think that would be better. The build system changes unfortunately made the old docs rather obsolete.

jonathansick commented on setup.cfg:
```diff
@@ -1,12 +1,82 @@
+package_dir =
+    = src
+packages = find:
+python_requires = >=3.9
```
3.8?

rra replied:
Yeah, I hadn't thought that through. Now fixed.

rra committed:
Relax Python dependency to allow 3.8 for now

This is a library and some of its clients have not yet been updated,
so be conservative in Python requirements for the time being.

jonathansick commented on docs/index.rst:
```diff
@@ -14,6 +14,15 @@ Templatekit is available from `PyPI <https://pypi.org/project/templatekit/>`_:
 
    pip install templatekit
 
+Getting started
+===============
+
+.. toctree::
+   :maxdepth: 1
+
+   changelog
```
Given there isn't a role for a "Getting started" in the way these docs are structured, the addition of a "Getting started" section here looks really out of place.

I'd suggest rolling it into the installation section until I can think of something better?

```suggestion

:doc:`See Templatekit's release history. <changelog>`_.

.. toctree::
   :hidden:
   
   changelog
```

jonathansick commented on setup.cfg:
```diff
@@ -1,12 +1,83 @@
-  .eggs
+[metadata]
+name = templatekit
+description = Tookit for rendering LSST project templates.
```
```suggestion
description = Tookit for rendering Rubin Observatory project templates.
```

jonathansick commented on setup.cfg:
```diff
@@ -1,12 +1,83 @@
+
+[options.extras_require]
+dev =
+    coverage[toml]==6.3.2
+    mypy==0.931
+    pre-commit==2.17.0
+    pytest==7.0.1
+    pytest-cov==3.0.0
```
I almost wonder if it's better to un-pin these and instead have a weekly cron job in GitHub Actions that tests again the latest dependencies? This might reduce some unnecessary PR churn from dependabot.

rra replied:
Yeah, good idea. Let me do that, and then I'll do it for Safir as well.

jonathansick reviewed (approved):
I've got some concerns about how the docs got restructured, but other than that, thanks for the overhaul of Templatekit.

rra committed:
Change LSST to Rubin Observatory

Co-authored-by: Jonathan Sick <jsick@lsst.org>

rra committed:
Incorporate changelog into installation section

Co-authored-by: Jonathan Sick <jsick@lsst.org>

rra committed:
Unpin dev dependencies, run CI from cron

Rather than forcing dependabot to constantly upgrade things, unpin
all of the dev dependencies and instead run the CI pipeline weekly
from cron so that we know if the new dev dependencies cuase failures.

rra committed:
Add release date for 0.5.0
